### PR TITLE
pam_unix: Report unusable hashes found by checksalt to syslog.

### PR DESCRIPTION
**Pull request by besser82**
libxcrypt can be build-time configured to support (or not support) various hashing methods.  Future versions will also have support for runtime configuration by the system's vendor and/or administrator.

For that reason adminstrator should be notified by pam if users cannot log into their account anymore because of such a change in the system's configuration of libxcrypt.

Also check for malformed hashes, like descrypt hashes starting with "$2...", which might have been generated by unsafe base64 encoding functions as used in glibc <= 2.16.
Such hashes are likely to be rejected by many recent implementations of libcrypt.

* modules/pam_unix/passverify.c (verify_pwd_hash): Report unusable hashes found by checksalt to syslog.

### CODE REVIEW

File: modules/pam_unix/passverify.c [hunk 1]
```diff
@@ -103,6 +103,40 @@ verify_pwd_hash(const char *p, char *hash, unsigned int nullok)
...
+			 * Check for malformed hashes, like descrypt hashes
+			 * starting with "$2...", which might have been
+			 * generated by unsafe base64 encoding functions
+			 * as used in glibc <= 2.16.
+			 * Such hashes are likely to be rejected by many
+			 * recent implementations of libcrypt.
+			 */
+			if (crypt_checksalt(hash) == CRYPT_SALT_INVALID) {
+				helper_log_err(LOG_ERR,
+				  "pam_unix(verify_pwd_hash): The setting "
```
Comment (t8m) [inline modules/pam_unix/passverify.c]: s/The setting used in the hash does not name/The hash does not use/

File: modules/pam_unix/passverify.c [hunk 2]
```diff
@@ -103,6 +103,40 @@ verify_pwd_hash(const char *p, char *hash, unsigned int nullok)
 			 * Ok, we don't know the crypt algorithm, but maybe
 			 * libcrypt knows about it? We should try it.
 			 */
+#if defined(CRYPT_CHECKSALT_AVAILABLE) && CRYPT_CHECKSALT_AVAILABLE
+			/*
+			 * Check for hashing methods that are disabled by
+			 * libcrypt configuration and/or system preset.
+			 */
+			if (crypt_checksalt(hash) == CRYPT_SALT_METHOD_DISABLED) {
```
Comment (t8m) [inline modules/pam_unix/passverify.c]: Please call the crypt_checksalt just once and assign the result to a variable and then test it. Also do not write the whole hash to the system log - something like hash \"%.3s...\" should be fine.

File: modules/pam_unix/passverify.c (hunk 1 above)
Comment (besser82) [reply modules/pam_unix/passverify.c]: Fixed.

File: modules/pam_unix/passverify.c (hunk 2 above)
Comment (besser82) [reply modules/pam_unix/passverify.c]: Fixed.

**Comment by besser82** [comment]
Should be fine now.

File: modules/pam_unix/passverify.c [hunk 3]
```diff
@@ -103,6 +103,42 @@ verify_pwd_hash(const char *p, char *hash, unsigned int nullok)
...
+			 * libcrypt configuration and/or system preset.
+			 */
+			if (retval_checksalt == CRYPT_SALT_METHOD_DISABLED) {
+				/*
+				 * pam_syslog() needs a pam handle,
+				 * but that's not available here.
+				 */
+				helper_log_err(LOG_ERR,
+				  "pam_unix(verify_pwd_hash): The method "
+				  "for computing the hash "\"%.6s\" has been "
```
Comment (t8m) [inline modules/pam_unix/passverify.c]: There seems to be extra " character. Did you try to build it?

Comment (besser82) [reply modules/pam_unix/passverify.c]: Fixed now.  Yes, I tried to build it, but that was before I made some changes to the log strings, which likely have introduced the extra `"`.

**Comment by t8m** [comment]
Merged, thank you.

**Pull request merged**